xwiki-platform-core/xwiki-platform-mentions/xwiki-platform-mentions-default/src/main/java/org/xwiki/mentions/internal/analyzer/CreatedDocumentMentionsAnalyzer.java

```diff
@@ -0,0 +1,170 @@
+ * This is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of
+ * the License, or (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+package org.xwiki.mentions.internal.analyzer;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+import javax.inject.Inject;
+import javax.inject.Singleton;
+
+import org.apache.commons.lang3.StringUtils;
+import org.xwiki.component.annotation.Component;
+import org.xwiki.mentions.MentionLocation;
+import org.xwiki.mentions.internal.MentionXDOMService;
+import org.xwiki.mentions.internal.MentionedActorReference;
+import org.xwiki.mentions.notifications.MentionNotificationParameter;
+import org.xwiki.mentions.notifications.MentionNotificationParameters;
+import org.xwiki.model.reference.DocumentReference;
+import org.xwiki.model.reference.EntityReference;
+import org.xwiki.rendering.block.MacroBlock;
+import org.xwiki.rendering.block.XDOM;
+import org.xwiki.rendering.syntax.Syntax;
+
+import com.xpn.xwiki.doc.XWikiDocument;
+import com.xpn.xwiki.objects.BaseObject;
+import com.xpn.xwiki.objects.LargeStringProperty;
+
+import static org.xwiki.mentions.MentionLocation.AWM_FIELD;
+import static org.xwiki.mentions.MentionLocation.DOCUMENT;
+
+/**
+ * Analyzes the new mentions on newly created documents.
+ *
+ * @version $Id$
+ * @since 12.10RC1
+ */
+@Component(roles = { CreatedDocumentMentionsAnalyzer.class })
+@Singleton
+public class CreatedDocumentMentionsAnalyzer extends AbstractDocumentMentionsAnalyzer
+{
+    @Inject
+    private MentionXDOMService xdomService;
+
+    /**
+     * Analyze a newly created document to identify new mentions.
+     *
+     * @param doc the analyzed document
+     * @param documentReference the reference of the document
+     * @param version the version of the created document
+     * @param authorReference the reference of the author of the document
+     * @return the list of {@link MentionNotificationParameters}, holding the new mentions identified for each entity of
+     *     the document.
+     */
+    public List<MentionNotificationParameters> analyze(XWikiDocument doc,
+        DocumentReference documentReference, String version, String authorReference)
+    {
+        Syntax syntax = doc.getSyntax();
+        List<MentionNotificationParameters> ret = new ArrayList<>();
```

Comment:
Note that you could simplify a lot the code of those analyzers, by considering they are not singleton: if you're using a `ComponentInstantiationStrategy.PER_LOOKUP` and you explicitely call the component manager to initialize them in the consumer, then you can consider that `ret` (bad name btw ;) ) is a field on those instances. And you don't have to pass it on each method. I don't know much how it will impact the memory footprint since I'm not sure if the consumer can handle several job at once. I guess that if the pool of thread is limited it shouldn't be an issue, you should have at most the same number of analyzer instance than the size of the pool of threads for the consumer.

Reply:
`ret` is something we can discuss, bit I often call the variable that will be returned by the current method like this.
For the rest, I'm not convinced trading an encapsulated aggregation of values (`ret` is not passed around) for a PER_LOOKUP strategy is a good thing

Reply:
> For the rest, I'm not convinced trading an encapsulated aggregation of values (ret is not passed around) for a PER_LOOKUP strategy is a good thing

why not? That's the perf impact that bothers you here or another reason? I'm not entirely against keeping it like that, it just that I think it would be easier to maintain with the `ret` considered as a field of the class.